src/pallene/coder.lua

```diff
@@ -1648,130 +1673,94 @@ end
-        while (1) {
-            ${body}
-        }
+        ${macro}($i, $cond, $iter, $count, $start, $limit, $step)
     ]], {
-        body = body
+        macro = macro,
+        i     = self:c_var(cmd.dst_i),
+        cond  = self:c_var(cmd.dst_cond),
+        iter  = self:c_var(cmd.dst_iter),
+        count = self:c_var(cmd.dst_count),
+        start = self:c_value(cmd.src_start),
+        limit = self:c_value(cmd.src_limit),
+        step  = self:c_value(cmd.src_step),
     }))
 end
 
-gen_cmd["For"] = function(self, cmd, func)
-    local typ = func.vars[cmd.dst].typ
+gen_cmd["ForStep"] = function(self, cmd, func)
+    local typ = func.vars[cmd.dst_i].typ
 
     local macro
     if     typ._tag == "types.T.Integer" then
-        macro = "PALLENE_INT_FOR_LOOP"
+        macro = "PALLENE_INT_FOR_STEP"
     elseif typ._tag == "types.T.Float" then
-        macro = "PALLENE_FLT_FOR_LOOP"
+        macro = "PALLENE_FLT_FOR_STEP"
     else
         tagged_union.error(typ._tag)
     end
 
     return (util.render([[
-        ${macro}_BEGIN($x, $start, $limit, $step)
-        {
-            $body
-        }
-        ${macro}_END
+        ${macro}($i, $cond, $iter, $count, $start, $limit, $step)
     ]], {
         macro = macro,
-        x     = self:c_var(cmd.dst),
+        i     = self:c_var(cmd.dst_i),
+        cond  = self:c_var(cmd.dst_cond),
+        iter  = self:c_var(cmd.dst_iter),
+        count = self:c_var(cmd.dst_count),
         start = self:c_value(cmd.src_start),
         limit = self:c_value(cmd.src_limit),
         step  = self:c_value(cmd.src_step),
-        body  = self:generate_cmd(func, cmd.body)
     }))
 end
 
+gen_cmd["Jmp"] = function(self, cmd, _func)
+    return "goto " .. self:c_label(cmd.target) .. ";"
+end
+
+gen_cmd["JmpIf"] = function(self, cmd, _func)
+    return util.render("if($v) {goto $t;} else {goto $f;}", {
+        v = self:c_value(cmd.src_cond),
+        t = self:c_label(cmd.target_true),
+        f = self:c_label(cmd.target_false),
+    })
+end
+
 gen_cmd["CheckGC"] = function(self, cmd, func)
     return util.render([[ luaC_condGC(L, ${update_stack_top}, (void)0); ]], {
         update_stack_top = self:update_stack_top(func, cmd) })
 end
 
+function Coder:generate_blocks(func)
+    local out = {}
+    for block_id,block in ipairs(func.blocks) do
+        -- putting "(void)0" at the last label so the C compiler doesn't complain about dangling
+        -- labels
+        local void_str = (block_id == #func.blocks) and "(void)0;" or ""
+        table.insert(out, util.render("$label:$void_str\n", {
+                label = self:c_label(block_id),
+                void_str = void_str,
+        }))
+        for _,cmd in ipairs(block.cmds) do
+            if cmd._tag ~= "ir.Cmd.Jmp" or cmd.target ~= block_id + 1 then
+                local cmd_str = self:generate_cmd(func, cmd) .. "\n"
+                table.insert(out, cmd_str)
+            end
+        end
+    end
+    return table.concat(out)
+end
+
```

Comment:
The `B:(void)0` looks weirdly indented in the generated C code, because our indenter does not recognize it as a label. What about add a line break before the return, and include a return statement for the void case too?

Non-void:

```
B3:
  return x1;
```

void:

```
B3:
  return;
```

Reply:
done!